Audio input  & AIC3204, PCM5102

ciutadellla opened the pull request:
Hi, I’ve been working on audio input for the Custom Board and wanted to create a pull request in case anyone likes to use or modify this work.  I originally had everything ADOC for a RP2350b + TLV AIC3204 but I made the effort to make it compatible with other codecs so its easy to add new ones in the future.

To enable simultaneous audio input and output, I lowered SAMPLE_RATE/BLOCK_SAMPLES. It has been tested with multiple stacked effects and works with AIC3204 and PCM5102.

You can try it out in the examples folder.

Best regards, make it shorter

    - Multiple codec AIC3204/PCM5102
    - Input and Output
    - Codec examples
    - Platformio to test Arduino Examples

ciutadellla committed:
## New Features:
    - Multi codec AIC3204/PCM5102
    - Input and Output
    - Codec examples
    - PiFosti Board Example
    - Platformio to test Arduino Examples

ciutadellla committed:
Remove comments

ciutadellla committed:
Removed Board Example

ghostintranslation commented:
Hi,

Before I dive into the code, can you explain what's the difference of AIC3204 vs PCM5102 other than it has inputs? Why the need of a separate AudioOutputI2S? I thought I2S is one standard the same for all I2S chips?

Also do you mean you changed SAMPLE_RATE and BLOCK_SAMPLES globally or only when using AIC3204?

And last I quickly see you used a PIO, but I think you only added the compiled version, not the source file? ("This file is autogenerated by pioasm; do not edit!")

Thanks for the work still, I will need to take time to read it all.

ciutadellla commented:
Hi, thanks for checking this out.

Let me clarify:

**PCM5102** (output only) is software-driven (CPU-based) I²S and no synchronization needed.
**AIC3204** (input + output) needs precise synchronization, initialization over I²C, handling of both input and output. hardware-driven (PIO/DMA) keeps input and output aligned (software-driven version cannot guarantee) which is why I added a separate AudioOutputI2S implementation tied to DMA + callback.

The SAMPLE_RATE and BLOCK_SAMPLES changend only when using **AIC3204**, I moved it to enums.h as it was needed to be shared in the hardware-driven aproach.

About he PIO file, you're right I forgot to commit it as it is working without it.

makingsoundmachines commented:
@ghostintranslation all this is based on a dev board we created for a small audio hack camp we did in Spain this summer; it was thought to be a fun exercise in C++ with the pico library, but then spontaneously evolved into making this work in Arduino land and with your Lib. do you want a board? it would make this more fun and less theoretical.

ghostintranslation commented:
Hey,
It's nice to see this project being used.
It would actually be nice to be able to test this code for real yes, send me a message on instagram we can talk about shipping there thanks.

I'll need time to review because of everything else I do, but I do see any PRs or messages and issues updates.

Thanks

ghostintranslation commented:
> About he PIO file, you're right I forgot to commit it as it is working without it.

@ciutadellla Can you commit the source PIO file?

Also, I'm wondering, is there really anything specific to the AIC3204 in the AudioInputI2S.h and in the PIO files actually? I didn't really see for example any I2C that would be specific to that codec, if it's only I2S it might be working with any other I2S chip?

ciutadellla commented:
@ghostintranslation Long time no see. The PIO is generic I2S; the I2C is the only part locked to the AIC3204. I Just uploaded the [PIO file ](https://github.com/ciutadellla/pico-audio/blob/audio-input/src/codec/AIC3204/audio_i2s.pio)

ciutadellla committed:
audio_i2s.pio

audio_i2s.pio

ghostintranslation commented:
Hey,
I know more about the Pico now, I will take some time to review this again thanks.

ghostintranslation commented on src/codec/AIC3204/AudioInputI2S.h:
```diff
@@ -0,0 +1,92 @@
+extern "C" {
+    #include "hardware/pio.h"
+    #include "hardware/dma.h"
+    #include "hardware/i2c.h"
```
I don't see I2C being used in this file? Same in AudioOutputI2S? Why including it here as it is already included in aic3204.h where it's being used?

ghostintranslation commented on src/enums.h:
```diff
@@ -0,0 +1,22 @@
+#define BLOCK_SIZE (AUDIO_BUFFER_SAMPLES * 2) // stereo interleaved
+
+#ifndef AUDIO_SAMPLE_RATE
+#define AUDIO_SAMPLE_RATE 16000//44100
```
This should remain 44100 at minimum.

ghostintranslation commented on src/enums.h:
```diff
@@ -0,0 +1,22 @@
+#define ENUMS_H
+
+#ifndef AUDIO_BLOCK_SAMPLES
+#define AUDIO_BLOCK_SAMPLES 64//128
```
Not sure this is better? Do you find a big difference with latency?
